_posts/2020-06-05-mirror-maker-2-eventhub.md

```diff
@@ -0,0 +1,290 @@
+It turns out that it's really important to set the configuration of `connections.max.idle.ms` and `metadata.max.age.ms`, for the producer, with a value less than 4 minutes!
+You could ask ... from where does this "magic" value come ?
+This is related to the behaviour of the Azure load balancers, in front of the Eveng Hub, which have an idle timeout setting of 4 minutes to 30 minutes.
+By default, it is set to 4 minutes.
+If a period of inactivity is longer than the timeout value, there's no guarantee that the TCP ession is maintained between the client and the cloud service.
+You can find more information on the official Microsoft [documentation](https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-tcp-idle-timeout#tcp-idle-timeout).
+
+The recommended configuration parameters for connecting to Azure Event Hub are also described [here](https://github.com/Azure/azure-event-hubs-for-kafka/blob/master/CONFIGURATION.md).
+
+The `mirrors` section describes what has to be mirrored and what should be the flow.
+In this case, the local "my-cluster" is the source and the "eventhub" is the target.
+
+Kafka Mirror Maker 2 is based on Kafka Connect and it uses a bunch of connectors for doing the mirroring, the heartbeating and the checkpointing.
+The `sourceConnector` is a specific Kafka Connect connector actually doing the mirroring, so reading from the local source cluster and making the messages available to Kafka Connect in order to mirror them to Azure Event Hub.
+The `heartbeatConnector` periodically checks connectivity between clusters.
+The `checkpointConnector` tracks and maps offsets for specified consumer groups using an offset sync topic and checkpoint topic.
+
+Create the `KafkaMirrorMaker2` on the Kubernetes cluster saving the above resource into a `kafka-mirror-maker-2-to-eh.yaml` file.
+The Strimzi cluster operator takes care of it, deploying Kafka Mirror Maker 2 using the above configuration.
+
+```shell
+kubectl apply -f kafka-mirror-maker-2-to-eh.yaml -n kafka
+```
+
+### Produce, mirror and consume!
+
+For the purpose of this demo, the application consuming the mirrored messages from Azure Event Hub is the `kafka-console-consumer` command line tool, provided with Apache Kafka, which needs a proper configuration as described below.
+
+```
+bootstrap.servers=<eventhubs-namespace>.servicebus.windows.net:9093
+security.protocol=SASL_SSL
+sasl.mechanism=PLAIN
+sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username="$ConnectionString" password="Endpoint=sb://<eventhubs-namespace>.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=<access-key>";
+```
+
+Save the above properties into a `kafka_eventhub.properties` file and start the consumer as following.
+
+```shell
+bin/kafka-console-consumer.sh --bootstrap-server <eventhubs-namespace>.servicebus.windows.net:9093 --topic testeh --consumer.config kafka_eventhub.properties
```

Comment:
Should you set the timout / idle configuration values here as well?

Reply:
Yes, fixed.